File: platform/polyglot/fetch_http_cache_response/py/fetch_http_cache_response/logger.py
"""Logger module for fetch_http_cache_response."""

import os
from typing import Callable, Optional, Protocol

DEBUG = 10
INFO = 20
WARN = 30
ERROR = 40

_LOG_LEVEL = int(os.getenv("FETCH_CACHE_LOG_LEVEL", str(DEBUG)))

PACKAGE_NAME = "fetch_http_cache_response"


class LoggerProtocol(Protocol):
    def debug(self, message: str) -> None: ...
    def info(self, message: str) -> None: ...
    def warn(self, message: str) -> None: ...
    def error(self, message: str) -> None: ...
    def child(self, instance_name: str) -> "LoggerProtocol": ...


class Logger:
    def __init__(
        self,
        package_name: str,
        filename: str,
        instance: str | None = None,
        level: int = _LOG_LEVEL,
        output: Callable[[str], None] | None = None,
    ):
        self._package = package_name
        self._filename = self._extract_filename(filename)
        self._instance = instance
        self._level = level
        self._output = output or print

    def _extract_filename(self, filename: str) -> str:
        if "." in filename:
            filename = filename.rsplit(".", 1)[-1]
        if filename.endswith(".py"):
            filename = filename[:-3]
        return filename

    def _log(self, level: str, level_value: int, message: str) -> None:
        if level_value >= self._level:
            if self._instance:
                prefix = f"[{level}] [{self._package}:{self._filename}] [{self._instance}]"
            else:
                prefix = f"[{level}] [{self._package}:{self._filename}]"
            self._output(f"{prefix} {message}")

    def debug(self, message: str) -> None:
        self._log("DEBUG", DEBUG, message)

    def info(self, message: str) -> None:
        self._log("INFO", INFO, message)

    def warn(self, message: str) -> None:
        self._log("WARN", WARN, message)

    def error(self, message: str) -> None:
        self._log("ERROR", ERROR, message)

    def child(self, instance_name: str) -> "Logger":
        return Logger(
            package_name=self._package,
            filename=f"{self._filename}:{self._instance or ''}",
            instance=instance_name,
            level=self._level,
            output=self._output,
        )


def create(
    filename: str,
    level: Optional[int] = None,
    output: Optional[Callable[[str], None]] = None,
) -> Logger:
    return Logger(
        package_name=PACKAGE_NAME,
        filename=filename,
        instance=None,
        level=level if level is not None else _LOG_LEVEL,
        output=output,
    )

File: platform/polyglot/fetch_http_cache_response/mjs/src/logger.ts
/**
 * Logger module for fetch-http-cache-response.
 */

const DEBUG = 10;
const INFO = 20;
const WARN = 30;
const ERROR = 40;

const LOG_LEVEL = parseInt(process.env.FETCH_CACHE_LOG_LEVEL || String(DEBUG), 10);

const PACKAGE_NAME = "fetch-http-cache-response";

function extractFilename(filename: string): string {
  if (filename.startsWith("file://")) filename = filename.split("/").pop()!;
  if (filename.includes("/")) filename = filename.split("/").pop()!;
  if (filename.endsWith(".ts")) filename = filename.slice(0, -3);
  if (filename.endsWith(".mjs")) filename = filename.slice(0, -4);
  if (filename.endsWith(".js")) filename = filename.slice(0, -3);
  return filename;
}

export interface LoggerLike {
  debug(message: string): void;
  info(message: string): void;
  warn(message: string): void;
  error(message: string): void;
  child(instanceName: string): LoggerLike;
}

export class Logger implements LoggerLike {
  #package: string;
  #filename: string;
  #instance: string | null;
  #level: number;
  #output: (msg: string) => void;

  constructor(
    packageName: string,
    filename: string,
    instance: string | null = null,
    level: number = LOG_LEVEL,
    output: (msg: string) => void = console.log,
  ) {
    this.#package = packageName;
    this.#filename = extractFilename(filename);
    this.#instance = instance;
    this.#level = level;
    this.#output = output;
  }

  #log(level: string, levelValue: number, message: string): void {
    if (levelValue >= this.#level) {
      const prefix = this.#instance
        ? `[${level}] [${this.#package}:${this.#filename}] [${this.#instance}]`
        : `[${level}] [${this.#package}:${this.#filename}]`;
      this.#output(`${prefix} ${message}`);
    }
  }

  debug(message: string): void { this.#log("DEBUG", DEBUG, message); }
  info(message: string): void { this.#log("INFO", INFO, message); }
  warn(message: string): void { this.#log("WARN", WARN, message); }
  error(message: string): void { this.#log("ERROR", ERROR, message); }

  child(instanceName: string): Logger {
    return new Logger(
      this.#package,
      `${this.#filename}:${this.#instance || ""}`,
      instanceName,
      this.#level,
      this.#output,
    );
  }
}

export function createLogger(
  filename: string,
  options: { level?: number; output?: (msg: string) => void } = {},
): Logger {
  return new Logger(
    PACKAGE_NAME,
    filename,
    null,
    options.level ?? LOG_LEVEL,
    options.output ?? console.log,
  );
}
